src/feature/hs/hs_circuit.c

```diff
@@ -1205,6 +1206,15 @@ hs_circ_cleanup(circuit_t *circ)
 {
   tor_assert(circ);
 
+  /* v2 specific circuits. */
```

Comment:
I'm fully onboard with this refactoring, however I'd like to make sure that it's indeed a no-behavior-change refactoring as the commit message claims.

In particular, this moves a whole bunch of code that used to reside in `circuit_about_to_free()` to `hs_circ_cleanup()`. What's the relationship of those two functions? Do they always get called at the same time?

Can this be documented in the commit message to clarify that no behavior change is done indeed?

Reply:
Actually, it does change behavior at this commit which another commit fixes.

I was mistaken at first when I did that one that `hs_circ_cleanup()` was only called for the circuit free() but it is also called for the close().

One approach would be to change the commit structure to add the distinction between `close` and `free` that e3f2df295fd98f1afdd285a4bdd185145d7c73b0 introduces. And then do this commit... We are talking some refactoring :).

Reply:
I did this work in the new PR. We now have a on_free, on_close and on_purpose functions. Will allow us to be much more future proof.